src/metabase/db/liquibase.clj

```diff
@@ -181,20 +189,44 @@
   see https://github.com/metabase/metabase/issues/3715"
   [db-type :- s/Keyword
    conn    :- java.sql.Connection]
-  (let [liquibase-table-name (if (#{:h2 :mysql} db-type)
-                               "DATABASECHANGELOG"
-                               "databasechangelog")
+  (let [liquibase-table-name (changelog-table-name db-type)
         fresh-install?       (jdbc/with-db-metadata [meta {:connection conn}] ; don't migrate on fresh install
                                (empty? (jdbc/metadata-query
                                         (.getTables meta nil nil liquibase-table-name (u/varargs String ["TABLE"])))))
         statement            (format "UPDATE %s SET FILENAME = ?" liquibase-table-name)]
     (when-not fresh-install?
       (jdbc/execute! {:connection conn} [statement "migrations/000_migrations.yaml"]))))
 
-(defn rollback-one
-  "Roll back the last migration."
-  [^Liquibase liquibase]
-  (.rollback liquibase 1 ""))
+(defn- extract-numbers
+  "Returns contiguous integers parsed from string s"
+  [s]
+  (map #(Integer/parseInt %) (re-seq #"\d+" s)))
+
+(defn- current-major-version
+  "Returns the major version of the running Metabase JAR"
+  []
+  (second (extract-numbers (:tag config/mb-version-info))))
+
+(defn rollback-major-version
+  "Roll back migrations later than given Metabase major version"
+  ;; default rollback to previous version
+  ([db-type conn liquibase]
+   ;; get current major version of Metabase we are running
+   (rollback-major-version db-type conn liquibase (dec (current-major-version))))
+
+  ;; with explicit target version
+  ([db-type conn ^Liquibase liquibase target-version]
+   (when (or (not (integer? target-version)) (< target-version 44))
+     (throw (IllegalArgumentException.
+             (trs "target version must be a number between 44 and the previous major version ({0}), inclusive" (current-major-version)))))
+   ;; count and rollback only the applied change set ids which come after the target version (only the "v..." IDs need to be considered)
+   (let [changeset-query (format "SELECT id FROM %s WHERE id LIKE 'v%%' ORDER BY ORDEREXECUTED ASC"
```

Comment:
It might not be the best idea in the whole universe to order by when things were executed... it's totally possible that things get executed out of order, for example if I am running some branch locally where I add `v45.00-100` and then I switch back to master and it has `v45.00-099` which was missing from my branch... in that case `-100` would have been executed before `-099`. 

Probably doesn't matter in that case but we do want to leave the option open to add new migrations to minor version releases of previous major version releases, which is why our ID format is `v<major>.<minor>-<migration>` as opposed to `v<major>-<migration>`. We might add a `v44.01-001` for some reason or another -- we've done it once in the past and we might have to do it again for a bug or security fix.

`ORDER BY id` should work tho.

Reply:
Unfortunately Liquibase does the rollback strictly sequentially by order applied, and not by ID, so we have to order by `orderexecuted` to get the same sequence Liquibase will see when it rolls back. You're right that could potentially cause some weird issues in a dev environment, but only if someone is using the rollback command, so it should not be an issue.